src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,395 @@
+            }
+
+            return GetArgumentExpression(formatArgumentSyntax);
+        }
+
+        protected static IMethodSymbol TryGetValidFormatMethodSymbol(
+            SyntaxNodeAnalysisContext context,
+            int numberOfArguments,
+            SymbolInfo symbolInfo)
+        {
+
+            if (symbolInfo.Symbol == null)
+            {
+                return null;
+            }
+
+            if (symbolInfo.Symbol.Kind != SymbolKind.Method)
+            {
+                return null;
+            }
+
+            var containingType = (INamedTypeSymbol)symbolInfo.Symbol.ContainingSymbol;
+
+            if (containingType.SpecialType != SpecialType.System_String)
+            {
+                return null;
+            }
+
+            var method = (IMethodSymbol)symbolInfo.Symbol;
+
+            if (method.Arity > 0)
+            {
+                return null;
+            }
+
+            return method;
+        }
+
+        private bool FormatCallWorksAtRuntime(string formatString, int numberOfPlaceholderArguments)
+        {
+            var testArray = new object[numberOfPlaceholderArguments];
+            for (var i = 0; i < numberOfPlaceholderArguments; i++)
+            {
+                testArray[i] = "test";
+            }
+
+            try
+            {
+                string.Format(formatString, testArray);
+            }
+            catch
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        protected void ValidateAndReportDiagnostic(
+            SyntaxNodeAnalysisContext context,
+            int numberOfPlaceholderArguments,
+            string formatString,
+            int formatStringPosition)
+        {
+            var formatStringWithEscapedBracketsRemoved = RemoveEscapedBrackets(formatString);
+
+            var matches = s_extractPlaceholdersRegex.Matches(formatStringWithEscapedBracketsRemoved);
+            foreach (Match match in matches)
+            {
+                var textInsideBrackets = match.Groups[1].Value;
+
+                if (!PlaceholderIndexIsValid(textInsideBrackets, numberOfPlaceholderArguments))
+                {
+                    var invalidPlaceholderText = "{" + textInsideBrackets + "}";
+                    var invalidPlaceholderLocation = Location.Create(
+                        context.Node.SyntaxTree,
+                        new Text.TextSpan(
+                            formatStringPosition + match.Index, 
+                            invalidPlaceholderText.Length));
+                    var diagnostic = Diagnostic.Create(
+                        Rule,
+                        invalidPlaceholderLocation, 
+                        invalidPlaceholderText);
+                    context.ReportDiagnostic(diagnostic);
+                }
+            }
+        }
+
+        private static string RemoveEscapedBrackets(string formatString)
+            => s_removeEscapedBracketsRegex.Replace(formatString, "  ");
```

Comment:
oh cute.  so you're removing the escapedBrackts and replacing with spaces to keep locations the same?  If so, please comment this code to explain that that's waht you're doing.